Added support for NVidia NVML to query GPU power and temperature

philipheinisch opened the pull request:
Adds two new command line switches -np xx and -nt xx to query GPU power and temperature using NVidia NVML, the sampling rate is controlled by the value xx in ms. For NVML to work, the NVidia CUDA toolkit needs to be installed. Temperature and power and corresponding timestamps are written to the output HDF5 file. This feature is controlled by the 'USENVML' define, so it can still be used without the CUDA toolkit. It could be made more user friendly in the future, but it works.

philipheinisch committed:
Added NVML interface

Commit:
WIndows NVML fixes

ranocha commented on CMakeLists.txt:
```diff
@@ -4,6 +4,15 @@ cmake_minimum_required (VERSION 3.5)
 PROJECT(toolkitICL)
 
 
+option(USENVML "Use NVML" ON)
```
We must either set the default to `OFF` or disable this option explicitly for the CI tests where CUDA isn't installed.

Since OpenCL does not target CUDA exclusively, I would vote for an explicit opt-in for this option.

philipheinisch replied:
What about setting default to off, but enabling it automatically, if CUDA is detected?

ranocha replied:
That sounds good.

ranocha commented on src/main.cpp:
```diff
@@ -13,10 +13,189 @@
 
 using namespace std;
 
+#define USENVML
```
This will result in failures if the option `USENVML` is disabled at configuration/build time.

ranocha commented on src/hdf5_io.cpp:
```diff
@@ -313,6 +313,65 @@ template bool h5_write_buffer(char const* filename, char const* varname, cl_long
+
+// write a buffer to an HDF5 file using compression
+template<typename TYPE>
+bool h5_append_buffer(char const* filename, char const* varname, TYPE const* data, size_t size)
```
How is this function different from `h5_write_buffer<TYPE>`? Aren't they exactly the same?

philipheinisch replied:
Right now it is not. The idea behind it was to be able to append data to exisiting HDF5 entires. This could be important to write live data for example the current GPU power to the HDF5 files. Right now everything gets buffered in RAM and dumped at the end. This is just a stub, because for short execution times, RAM buffering is more efficient.

ranocha replied:
I would favor either a TODO note in the code (without implementation) or a full implementation. Right now, someone who reads the code will expect that this function can be used for the purpose of its name.

philipheinisch replied:
For now, RAM buffering seems fine, so I will remove the stub code and add a ToDo.

ranocha commented on src/CMakeLists.txt:
```diff
@@ -17,6 +17,13 @@ set(SOURCES main.cpp hdf5_io.cpp ocl_dev_mgr.cpp ${HEADER})
+if (MSVC)
+link_directories(${OpenCL_LIBRARIES} ${HDF5_HL_LIBRARIES} ${HDF5_LIBRARIES} ${CUDA_TOOLKIT_ROOT_DIR}/lib/x64)
+ELSE(MSVC)
+link_directories(toolkitICL ${OpenCL_LIBRARIES} ${HDF5_HL_LIBRARIES} ${HDF5_LIBRARIES} ${CUDA_TOOLKIT_ROOT_DIR}/lib/x86_64-linux-gnu/stubs/)
```
Why did you add `toolkitICL` here?

philipheinisch replied:
No specific reason, must ahve been a copy&paste mistake.

ranocha commented on src/main.cpp:
```diff
@@ -13,10 +13,189 @@
+
+		nvmlDeviceGetPowerUsage(device, &temp); 
+    //time(&rawtime);
+    gettimeofday(&rawtime, NULL);
```
This part looks weird because of mixed indentation (tabs & spaces).

philipheinisch replied:
It seemed like the easiest option to implement shared memory for the std::threads without having to define a specific class. I agree, that it is not the most elegant solution. Any better ideas?

philipheinisch replied:
It seems like changing between VS2017/VS2015 and VS Code has some drawbacks regarding auto indentations, whitespaces and so on.

ranocha commented on src/main.cpp:
```diff
@@ -13,10 +13,189 @@
+
+#if defined(USENVML)
+#include <nvml.h>
+bool nv_log_pwr = false;
```
I have some personal objections against global variables but that might be just because of their drawbacks in Julia :wink:

philipheinisch replied:
It seemed like the easiest option to implement shared memory for the std::threads without having to define a specific class. I agree, that it is not the most elegant solution. Any better ideas?

ranocha replied:
Local variables and a lambda/closure? You can also use global variables if you like. I won't reject your changes because of that :wink:

philipheinisch replied:
Maybe we could use global variables for now and if all works well, wrap everything in a seperate class with correct non global variables for shared memory when we have time for cleanup. It should be done at some point anyways, otherwise the main.cpp will get overcrowded, especially if we also implement things like  Intel® Power Gadget.

ranocha commented on src/main.cpp:
```diff
@@ -300,7 +499,7 @@ int main(int argc, char *argv[]) {
 
   for (string const& kernel_name : kernel_list) {
     exec_time = exec_time + dev_mgr.execute_kernelNA(*(dev_mgr.getKernelbyName(0, "ocl_Kernel", kernel_name)),
-                                                     dev_mgr.get_queue(0, 0), range_start, global_range, local_range);
```
Why did you delete the whitespace here? It is harder to read (at least for me) in this new version.

philipheinisch replied:
It seems like changing between VS2017/VS2015 and VS Code has some drawbacks regarding auto indentations, whitespaces and so on.

ranocha commented on src/main.cpp:
```diff
@@ -318,7 +517,56 @@ int main(int argc, char *argv[]) {
+
+#if defined(USENVML)
+  nv_log_pwr = false;
+  nv_log_tmp = false;
```
Could you please use a homogeneous indentation?

philipheinisch committed:
Added automatic CUDA detection

philipheinisch committed:
NVML define gets added automatically by CMake

philipheinisch commented:
Does it correctly find the CUDA toolkit on your system now @ranocha? And how can I re-run the CI checks?

philipheinisch committed:
Added NVML description to README.md

ranocha commented:
Yes, it seems to find CUDA but cmake errors:
```
toolkitICL/build$ cmake ..
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done                                            
-- Looking for pthread.h                                                            
-- Looking for pthread.h - found                                                    
-- Looking for pthread_create                                                       
-- Looking for pthread_create - not found                                           
-- Looking for pthread_create in pthreads                                           
-- Looking for pthread_create in pthreads - not found                               
-- Looking for pthread_create in pthread                                            
-- Looking for pthread_create in pthread - found                                    
-- Found Threads: TRUE                                                              
-- Found CUDA: /usr (found version "9.1")                                           
-- Using NVML                                                                       
CMake Error at CMakeLists.txt:11 (add_compile_definitions):                         
  Unknown CMake command "add_compile_definitions".                                  
                                                                                    
                                                                                    
-- Configuring incomplete, errors occurred!                                         
```

ranocha commented:
The CI tests should run automatically if you push new commits.

philipheinisch commented:
What is your cmake version? Maybe add_compile_definitions (https://cmake.org/cmake/help/v3.12/command/add_compile_definitions.html) is to new of a feature to be available on most platforms? What would be an acceptable minimal cmake version? 

Is it possible, that the CI test are not working anymore due to the merge conflicts?

ranocha commented:
> Is it possible, that the CI test are not working anymore due to the merge conflicts?

That would be a bit strange but I'm not sure.

ranocha commented:
> What is your cmake version?

I can have a look in the afternoon. At my computer at TU BS, I'm using `cmake version 3.7.2`.

philipheinisch committed:
Changed CMakeLists to support older version of Cmake

philipheinisch commented:
> I can have a look in the afternoon. At my computer at TU BS, I'm using `cmake version 3.7.2`.

That is indeed very old... But could you retry at your office PC? It should work now, even with ancient CMake versions.

ranocha commented:
> But could you retry at your office PC? It should work now, even with ancient CMake versions.

It can be compiled without errors (but some tests fail - I will have a look at that later).

philipheinisch commented:
Ok, if it compiles correctly on different platforms, I will try and fix the reaming indentation problems and than we can probably try to merge?

ranocha commented:
I still have to test it at home - I have no CUDA at my office PC.Afterwards, we can fix the remaining parts & merge conflicts. Then, CI tests will hopefully pass again.

ranocha commented:
The build using CUDA is fine if I run `cmake` (version 3.10.2). However, if I use `cmake -DUSENVML=OFF ..`, the build fails because CUDA should still be linked
```
CMakeFiles/toolkitICL.dir/main.cpp.o: In function `nv_log_pwr_func()':
main.cpp:(.text+0x11d0): undefined reference to `nvmlInit_v2'
main.cpp:(.text+0x11ef): undefined reference to `nvmlDeviceGetHandleByIndex_v2'
main.cpp:(.text+0x1273): undefined reference to `nvmlDeviceGetPowerUsage'
main.cpp:(.text+0x12d0): undefined reference to `nvmlShutdown'
CMakeFiles/toolkitICL.dir/main.cpp.o: In function `nv_log_tmp_func()':
main.cpp:(.text+0x1380): undefined reference to `nvmlInit_v2'
main.cpp:(.text+0x1390): undefined reference to `nvmlDeviceGetHandleByIndex_v2'
main.cpp:(.text+0x1427): undefined reference to `nvmlDeviceGetTemperature'
main.cpp:(.text+0x148a): undefined reference to `nvmlShutdown'
collect2: error: ld returned 1 exit status
src/CMakeFiles/toolkitICL.dir/build.make:156: recipe for target 'bin/toolkitICL' failed
make[2]: *** [bin/toolkitICL] Error 1
CMakeFiles/Makefile2:85: recipe for target 'src/CMakeFiles/toolkitICL.dir/all' failed
make[1]: *** [src/CMakeFiles/toolkitICL.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Edit: But that can also be attributed to a weird use case - or a dump user :wink:

ranocha commented on CMakeLists.txt:
```diff
@@ -3,6 +3,17 @@
+ MESSAGE(STATUS "Using NVML")
+ #add_compile_definitions(USENVML)
+ add_definitions(-DUSENVML)
+ELSEIF(CUDA_FOUND)
```
Shouldn't this be `ELSE`?

philipheinisch replied:
Yes, it should have been...

ranocha commented on CMakeLists.txt:
```diff
@@ -3,6 +3,17 @@
+ find_package(CUDA)
+ 
+IF( CUDA_FOUND)
+ option(USENVML "Use NVML" ON)
```
Maybe it would be better to use something like
```
IF(CUDA_FOUND)
  option(USENVML "Use NVML" ON)
ELSE()
  option(USENVML "Use NVML" OFF)
ENDIF()

IF(USENVML)
  MESSAGE(STATUS "Using NVML")
  add_definitions(-DUSENVML)
ENDIF()
```

philipheinisch replied:
Using Windows, this overrides 'cmake -DUSENVML=OFF' if CUDA is detected. I'm guessing this is the intention behind it?

ranocha replied:
You're right We (you?) should decide whether `USENVML` shall be an option for the user or not. If it is no option, we can use something as above. Otherwise, an additional check has to be added to the first `IF`. But I think it might be nice to turn `USENVML=OFF` if CUDA isn't found (with an additional warning, maybe?).

ranocha commented on CMakeLists.txt:
```diff
@@ -42,7 +56,7 @@ ENDIF(HDF5_FOUND)
 find_package(OpenCL REQUIRED)
 IF(OpenCL_FOUND)
-  MESSAGE(STATUS "Looking for OpenCL - found at\n   ${OpenCL_INCLUDE_DIRS}\n   ${OpenCL_LIBRARY}")
+  MESSAGE(STATUS "Looking for OpenCL - found at\n   ${OpenCL_INCLUDE_DIRS}\n  ${OpenCL_LIBRARY}")
```
Why did you remove the indentation here? Now, it looks a bit weird for me:
```
   /usr/include
  /usr/lib/x86_64-linux-gnu/libOpenCL.so
```

philipheinisch replied:
It wasn't intentional. VS2017 just likes to changed indentations.

ranocha commented on README.md:
```diff
@@ -9,13 +9,14 @@ to use cross platform tool to execute a set of kernels for example on compute cl
 - OpenCL (headers and drivers)
 - HDF5
 - CMake
+- CUDA Toolkit (only for GPU power/temperature logging)
```
Maybe a remark that CUDA will be detected automatically if present would be good? It doesn't need to be here but somewhere in the README would be good.

philipheinisch committed:
Added additional explanation to README.md and fixed CMakeLists.txt

philipheinisch commented:
I added an explanation to the readme, is this what you imagined? I also changed the CMake file to include 
> IF(CUDA_FOUND)
>   option(USENVML "Use NVML" ON)
> ELSE()
>   option(USENVML "Use NVML" OFF)
> ENDIF()
> 
> IF(USENVML)
>   MESSAGE(STATUS "Using NVML")
>   add_definitions(-DUSENVML)
> ENDIF()
> ```

this is likely the most flexible solution.

philipheinisch committed:
Fixed typo in README.md

ranocha committed:
minor cleanup before rebase

ranocha commented:
Finally, the tests pass :tada: Thanks, @philipheinisch, good work as always (with a debatable indentation style :wink:).
